capstone/backend/src/main/java/com/google/servlets/StoreRankingsServlet.java

```diff
@@ -0,0 +1,179 @@
+import com.google.cloud.spanner.SpannerException;
+import com.google.gson.Gson;
+import com.google.spanner.LibraryFunctions;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/api/v1/get-store-rankings")
+public class StoreRankingsServlet extends HttpServlet {
+  private static final String API_KEY = "INSERT_GOOGLE_API_KEY_HERE";
+  private static final double AVALIABLE_ITEMS_WEIGHT = 3;
+  private static final double DISTANCE_WEIGHT = -0.0005;
+  private static final double MILES_TO_METERS = 1609.34;
+  private static final double PRICE_WEIGHT = -1;
+  private Gson g = new Gson();
+
+  private class DistanceResponse {
+    private class Row {
+      private List<Element> elements;
+
+      private class Element {
+        private Distance distance;
+        private Duration duration;
+        private String status;
+
+        private class Distance {
+          String text;
+          String value;
+        }
+
+        private class Duration {
+          String text;
+          String value;
+        }
+      }
+    }
+
+    private String status;
+    private List<String> origin_addresses;
+    private List<String> destination_addresses;
+    private List<Row> rows;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserPreferences userPreferences = getUserPreferencesAndValidate(request);
+    if (userPreferences == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid request syntax.");
+      return;
+    }
+    List<Store> stores = new ArrayList<Store>();
+    try {
+      stores = getStores(userPreferences.getSelectedItemTypes());
+    } catch (SpannerException e) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage());
+      return;
+    }
+    Map<String, Integer> distances =
+        getDistances(
+            stores.stream().map(store -> store.getStoreAddress()).collect(Collectors.toList()),
+            userPreferences.getLocation());
+    stores =
+        stores.stream()
+            .filter(
+                store ->
+                    (distances.get(store.getStoreAddress())
+                        < (userPreferences.getDistancePreference() * MILES_TO_METERS)))
+            .collect(Collectors.toList());
+    rankStores(stores, distances);
+    response.setContentType("application/json");
+    response.setStatus(HttpServletResponse.SC_OK);
+    response.getWriter().println(g.toJson(stores));
+  }
+
+  private void rankStores(List<Store> stores, Map<String, Integer> distances) {
```

Comment:
Add a description of how the ranking algorithm works? or  something like `See [Link to doc]`

Reply:
Once I talk to Brett and Carolyn and we finalize the algorithm and weights we are going to use, I will link a document explaining how the ranking algorithm works.